blockstore.go

```diff
@@ -179,23 +180,54 @@ func (bs *blockstore) Put(ctx context.Context, block blocks.Block) error {
 	return bs.datastore.Put(ctx, k, block.RawData())
 }
 
-func (bs *blockstore) PutMany(ctx context.Context, blocks []blocks.Block) error {
+func (bs *blockstore) PutMany(ctx context.Context, blocksList []blocks.Block) error {
+	if len(blocksList) == 1 {
+		// performance fast-path
+		return bs.Put(ctx, blocksList[0])
+	}
+
 	t, err := bs.datastore.Batch(ctx)
 	if err != nil {
 		return err
 	}
-	for _, b := range blocks {
-		k := dshelp.MultihashToDsKey(b.Cid().Hash())
-		exists, err := bs.datastore.Has(ctx, k)
-		if err == nil && exists {
-			continue
-		}
 
-		err = t.Put(ctx, k, b.RawData())
-		if err != nil {
-			return err
-		}
+	jobs := make(chan blocks.Block)
+	g, ctx := errgroup.WithContext(ctx)
+
+	const maxWorkers = 5
+
+	numWorkers := maxWorkers
+	if len(blocksList) < numWorkers {
+		numWorkers = len(blocksList)
 	}
+
+	for i := 0; i < numWorkers; i++ {
+		g.Go(func() error {
+			for b := range jobs {
+				k := dshelp.MultihashToDsKey(b.Cid().Hash())
+				exists, err := bs.datastore.Has(ctx, k)
+				if err == nil && exists {
+					continue
+				}
+				err = t.Put(ctx, k, b.RawData())
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+	}
+
+	for _, b := range blocksList {
+		jobs <- b
+	}
```

Comment:
Will this deadlock if e.g. all the workers return an error, and there are more workers than elements in blocksList?